lib/Test2/Event/Faceted.pm

```diff
@@ -0,0 +1,221 @@
+use warnings;
+
+our $VERSION = '1.302077';
+
+BEGIN { require Test2::Event; our @ISA = qw(Test2::Event) }
+use Test2::Util::HashBase;
+
+use Scalar::Util();
+
+require Test2::EventFacet::Amnesty;
+require Test2::EventFacet::Assert;
+require Test2::EventFacet::Info;
+require Test2::EventFacet::Nest;
+require Test2::EventFacet::Plan;
+require Test2::EventFacet::Stop;
+require Test2::EventFacet::Trace;
+
+sub no_legacy_facets() { 1 }
+
+# Optimization
+my %MAKE_FACET = ( ARRAY => 1, HASH => 1 );
+my %AUTO_FACET = (
+    assert => 'Test2::EventFacet::Assert',
+    nest   => 'Test2::EventFacet::Nest',
+    plan   => 'Test2::EventFacet::Plan',
+    stop   => 'Test2::EventFacet::Stop',
+
+    # These get handled differently.
+    # info => 'Test2::EventFacet::info',
+    # info => 'Test2::EventFacet::amnesty',
+);
+
+sub init {
+    my $self = shift;
+
+    $self->{+_FACETS} ||= delete $self->{facets}
+        if $self->{facets};
+
+    $self->{+TRACE} ||= $self->{+_FACETS}->{trace}
+        if $self->{+_FACETS}->{trace};
+
+    for my $facet (keys %AUTO_FACET) {
+        my $raw = delete $self->{$facet} or next;
+        my $class = $AUTO_FACET{$facet};
+        my $ref = ref($raw);
+        $self->{+_FACETS}->{$facet} = $class->new($raw) and next
+            if $MAKE_FACET{$ref};
+
+        $self->{+_FACETS}->{$facet} = $raw and next
+            if $ref eq $class || Scalar::Util::blessed($raw) && $raw->isa($class);
+
+        Carp::croak("The '$facet' attribute takes an arrayref, a hashref, or an instance of the '$class' class");
+    }
+
+    # Normalize pass/fail to be a boolean.
+    $self->{+_FACETS}->{assert}->{pass} = $self->{+_FACETS}->{assert}->{pass} ? 1 : 0
+        if $self->{+_FACETS}->{assert};
+
+    if (my $info = delete $self->{info}) {
+        $self->add_info(@$info);
+    }
+
+    if (my $amnesty = delete $self->{amnesty}) {
+        $self->add_amnesty(@$amnesty);
+    }
+
+    # If diagnostics were provided put them in info, unless this is a passing
+    # assert. If this is not an assert we should also set the gravity to 100 so
+    # the diags can be seen. Do not mess with gravity if there is an assert,
+    # that is handled better in gravity().
+    if(my $diag = delete $self->{diag}) {
+        my $assert = $self->{+_FACETS}->{assert};
+        if (!$assert || !$assert->{pass}) {
+            $self->add_info([diag => $_]) for @$diag;
+        }
+
+        $self->{+GRAVITY} = 100 unless defined $self->{+GRAVITY} || $assert;
+    }
+
+    # Stick any notes into info.
+    if(my $note = delete $self->{note}) {
+        $self->add_info([note => $_]) for @$note;
+    }
+
+    1;
+}
+
+sub gravity {
+    my $self = shift;
+    return $self->{+GRAVITY} if defined $self->{+GRAVITY};
+    return 100 if $self->causes_fail;
+    return 0;
+}
+
+sub facets {
+    my $self = shift;
+
+    my %facets = $self->{+_FACETS} ? %{$self->{+_FACETS}} : ();
+
+    push @{$facets{amnesty}} => @{$self->{+_AMNESTY}} if $self->{+_AMNESTY};
```

Comment:
Won't this (and similar code elsewhere) push onto `$self->{+_FACETS}` over and over every time the `facets` method is called?